src/hotspot/share/cds/dynamicArchive.cpp

```diff
@@ -377,18 +377,35 @@ void DynamicArchive::check_for_dynamic_dump() {
   }
 }
 
-void DynamicArchive::prepare_for_dump_at_exit() {
-  EXCEPTION_MARK;
-  ResourceMark rm(THREAD);
-  MetaspaceShared::link_shared_classes(false/*not from jcmd*/, THREAD);
-  if (HAS_PENDING_EXCEPTION) {
-    log_error(cds)("Dynamic dump has failed");
-    log_error(cds)("%s: %s", PENDING_EXCEPTION->klass()->external_name(),
-                   java_lang_String::as_utf8_string(java_lang_Throwable::message(PENDING_EXCEPTION)));
-    // We cannot continue to dump the archive anymore.
-    DynamicDumpSharedSpaces = false;
-    CLEAR_PENDING_EXCEPTION;
+void DynamicArchive::dump_at_exit(JavaThread* current, const char* archive_name) {
+  ExceptionMark em(current);
+  ResourceMark rm(current);
+  HandleMark hm(current);
+
+  if (!(DynamicDumpSharedSpaces && archive_name != nullptr)) {
+    return;
+  }
+
+  log_info(cds,dynamic)("Preparing for dynamic dump at exit in thread %s", current->name());
+
+  MetaspaceShared::link_shared_classes(false/*not from jcmd*/, current);
```

Comment:
link_shared_classes is declared with TRAPS, so this should have a
JavaThread* THREAD = current;
and use HAS_PENDING_EXCEPTION, CLEAR_PENDING_EXCEPTION macros.

Reply:
Okay